File: sdks/typescript/src/request.ts
import { timestampFromDate } from "@bufbuild/protobuf/wkt";

import type { CommandContext } from "../../../protocols/generated/typescript/common/v1/command_context_pb.js";
import type { ClientConfig } from "./config.js";
import { validateAttempts, validateMetadata } from "./config.js";
import { MindcladeError } from "./error.js";
import type { PaginationLimits } from "./pagination.js";
import { platformMetadata } from "./platform.js";
import type { Runtime } from "./runtime.js";

export type {
	PaginationLimits,
	PaginationOptions,
	PaginationPage,
} from "./pagination.js";
export { paginate } from "./pagination.js";

/**
 * Explicit, named permission to retry an RPC the safety table classifies as
 * non-idempotent. There is deliberately no bare boolean: the only way to obtain
 * the token is {@link withUnsafeRetryOfNonIdempotent}, which forces the caller
 * to record why duplicate execution is acceptable for their call.
 */
export interface UnsafeRetryOfNonIdempotent {
	readonly justification: string;
	readonly acknowledged: true;
}

/**
 * Mints the named override that allows an `unsafe` RPC to be retried.
 * Routes pinned by `isNeverRetryable` ignore the token entirely.
 */
export const withUnsafeRetryOfNonIdempotent = (
	justification: string,
): UnsafeRetryOfNonIdempotent => {
	validateJustification(justification);
	return Object.freeze({ acknowledged: true as const, justification });
};

export interface SdkCallOptions {
	readonly requestId?: string;
	readonly traceId?: string;
	/** Total budget across every attempt, backoff, and credential acquisition. */
	readonly timeoutMs?: number;
	readonly signal?: AbortSignal;
	/** Per-request attempt ceiling, one to eight, overriding the client policy. */
	readonly maxAttempts?: number;
	/** Named permission to retry a non-idempotent RPC. Never a bare boolean. */
	readonly unsafeRetryOfNonIdempotent?: UnsafeRetryOfNonIdempotent;
	/** Raw fenced worker identity carried only as transport metadata. */
	readonly workerId?: string;
	/** Sensitive lease capability carried only as transport metadata. */
	readonly leaseToken?: string;
}

/** Per-attempt retry position advertised to the server on every attempt. */
export interface RetryAttemptState {
	/** Zero-based index of the attempt being issued. */
	readonly attempt: number;
	/** Milliseconds left in the caller's total budget when the attempt starts. */
	readonly remainingMs: number;
}

export interface SubmitOptions extends SdkCallOptions {
	readonly idempotencyKey: string;
	readonly correlationId?: string;
	readonly causationId?: string;
	readonly cancellationTokenId?: string;
}

export interface WaitOptions extends SdkCallOptions {
	readonly waitTimeoutMs?: number;
	readonly pollIntervalMs?: number;
}

/** Options accepted by every auto-paginating list method. */
export interface ListOptions extends SdkCallOptions {
	/** Page and item budgets for the transparent traversal. */
	readonly limits?: PaginationLimits | undefined;
}

export interface PreparedCall {
	readonly requestId: string;
	readonly traceId: string;
	readonly deadlineMs: number;
	readonly signal: AbortSignal | undefined;
	readonly maxAttempts: number | undefined;
	readonly unsafeRetryOfNonIdempotent: UnsafeRetryOfNonIdempotent | undefined;
	readonly workerId: string | undefined;
	readonly leaseToken: string | undefined;
}

export const prepareCall = (
	config: ClientConfig,
	runtime: Runtime,
	options: SdkCallOptions,
): PreparedCall => {
	const timeoutMs = options.timeoutMs ?? config.defaultTimeoutMs;
	validateDuration("call timeout", timeoutMs);
	const requestId = options.requestId ?? runtime.requestId();
	const traceId = options.traceId ?? requestId;
	validateMetadata("request ID", requestId, true);
	validateMetadata("trace ID", traceId, true);
	if (options.workerId !== undefined) validateMetadata("worker ID", options.workerId, true);
	if (options.leaseToken !== undefined) validateLeaseToken(options.leaseToken);
	if (options.maxAttempts !== undefined) validateAttempts("call max attempts", options.maxAttempts);
	if (options.unsafeRetryOfNonIdempotent !== undefined) {
		validateUnsafeRetryOverride(options.unsafeRetryOfNonIdempotent);
	}
	return {
		requestId,
		traceId,
		deadlineMs: runtime.nowMs() + timeoutMs,
		signal: options.signal,
		maxAttempts: options.maxAttempts,
		unsafeRetryOfNonIdempotent: options.unsafeRetryOfNonIdempotent,
		workerId: options.workerId,
		leaseToken: options.leaseToken,
	};
};

export const callHeaders = (
	config: ClientConfig,
	call: PreparedCall,
	retry: RetryAttemptState,
	idempotencyKey?: string,
): Headers => {
	const headers = new Headers({
		"x-mindclade-sdk": platformMetadata(config.omitPlatformMetadata),
		"x-mindclade-expected-tenant": config.identity.tenantId,
		"x-mindclade-expected-project": config.identity.projectId,
		"x-mindclade-expected-principal": config.identity.principalId,
		"x-request-id": call.requestId,
		"x-trace-id": call.traceId,
		"x-mindclade-retry-count": String(Math.max(0, Math.floor(retry.attempt))),
		"x-mindclade-timeout-ms": String(Math.max(1, Math.floor(retry.remainingMs))),
	});
	if (idempotencyKey !== undefined) {
		validateMetadata("idempotency key", idempotencyKey, true);
		headers.set("idempotency-key", idempotencyKey);
	}
	if (call.workerId !== undefined) headers.set("x-mindclade-worker-id", call.workerId);
	if (call.leaseToken !== undefined) headers.set("x-mindclade-lease-token", call.leaseToken);
	return headers;
};

export const commandContext = (
	config: ClientConfig,
	call: PreparedCall,
	options: SubmitOptions,
): CommandContext => {
	validateSubmitOptions(options);
	return {
		$typeName: "mindclade.common.v1.CommandContext",
		requestId: call.requestId,
		idempotencyKey: options.idempotencyKey,
		principalId: config.identity.principalId,
		traceId: call.traceId,
		deadline: timestampFromDate(new Date(call.deadlineMs)),
		// Canonical command identity is computed and materialized by the
		// control plane over the received generated message.
		canonicalRequestDigest: "",
		tenantId: config.identity.tenantId,
		projectId: config.identity.projectId,
		correlationId: checkedOptional("correlation ID", options.correlationId),
		causationId: checkedOptional("causation ID", options.causationId),
		cancellationTokenId: checkedOptional("cancellation token ID", options.cancellationTokenId),
	};
};

export const validateSubmitOptions = (options: SubmitOptions): void => {
	validateMetadata("idempotency key", options.idempotencyKey, true);
};

export const validateResource = (name: string, value: string): void =>
	validateMetadata(name, value, true);

export const validateDuration = (name: string, value: number): void => {
	if (!Number.isFinite(value) || value <= 0 || value > 86_400_000) {
		throw MindcladeError.invalidArgument(`${name} must be positive and at most twenty-four hours`);
	}
};

const checkedOptional = (name: string, value: string | undefined): string => {
	if (value === undefined) return "";
	validateMetadata(name, value, true);
	return value;
};

const validateUnsafeRetryOverride = (override: UnsafeRetryOfNonIdempotent): void => {
	if (override.acknowledged !== true) {
		throw MindcladeError.invalidArgument(
			"unsafe retry of a non-idempotent RPC must be acknowledged explicitly",
		);
	}
	validateJustification(override.justification);
};

/** Justifications are prose, so a single space is allowed where metadata forbids it. */
const validateJustification = (value: string): void => {
	if (
		value.length === 0 ||
		value.length > 256 ||
		[...value].some((character) => {
			const code = character.charCodeAt(0);
			return code < 0x20 || code > 0x7e;
		})
	) {
		throw MindcladeError.invalidArgument(
			"unsafe retry justification must contain one to 256 printable ASCII characters",
		);
	}
};

const validateLeaseToken = (value: string): void => {
	if (
		value.length === 0 ||
		value.length > 4096 ||
		[...value].some((character) => {
			const code = character.charCodeAt(0);
			return code < 0x21 || code > 0x7e;
		})
	) {
		throw MindcladeError.invalidArgument(
			"lease token must contain at most 4096 visible ASCII characters",
		);
	}
};

File: sdks/rust/src/request.rs
use std::{
    collections::{HashSet, VecDeque},
    fmt,
    future::Future,
    sync::atomic::{AtomicU64, Ordering},
    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
};

use prost_types::Timestamp;

use crate::{Config, Error, config::validate_metadata_value};

static REQUEST_SEQUENCE: AtomicU64 = AtomicU64::new(1);

const DEFAULT_PAGINATION_MAX_PAGES: usize = 100;
const DEFAULT_PAGINATION_MAX_ITEMS: usize = 10_000;
const HARD_PAGINATION_MAX_PAGES: usize = 1_000;
const HARD_PAGINATION_MAX_ITEMS: usize = 1_000_000;

/// Hard limits for lazy traversal of opaque-token list pages.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct PaginationLimits {
    max_pages: usize,
    max_items: usize,
}

impl PaginationLimits {
    /// Builds validated pagination limits.
    ///
    /// # Errors
    ///
    /// Returns an error for zero or unreasonably large limits.
    pub fn new(max_pages: usize, max_items: usize) -> Result<Self, Error> {
        if !(1..=HARD_PAGINATION_MAX_PAGES).contains(&max_pages) {
            return Err(Error::invalid_argument(
                "pagination max pages must be in [1, 1000]",
            ));
        }
        if !(1..=HARD_PAGINATION_MAX_ITEMS).contains(&max_items) {
            return Err(Error::invalid_argument(
                "pagination max items must be in [1, 1000000]",
            ));
        }
        Ok(Self {
            max_pages,
            max_items,
        })
    }
}

impl Default for PaginationLimits {
    fn default() -> Self {
        Self {
            max_pages: DEFAULT_PAGINATION_MAX_PAGES,
            max_items: DEFAULT_PAGINATION_MAX_ITEMS,
        }
    }
}

/// One generated-facade page adapted for [`Paginator`].
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PaginationPage<T> {
    pub items: Vec<T>,
    pub next_page_token: String,
}

impl<T> PaginationPage<T> {
    #[must_use]
    pub fn new(items: Vec<T>, next_page_token: impl Into<String>) -> Self {
        Self {
            items,
            next_page_token: next_page_token.into(),
        }
    }
}

/// Lazy, bounded automatic pagination over an ergonomic facade list method.
///
/// The fetch closure receives each opaque token without normalization. Use
/// [`Paginator::try_next`] in a loop; it rejects repeated cursors and reports
/// budget exhaustion rather than presenting a partial traversal as complete.
pub struct Paginator<T, Fetch> {
    fetch_page: Fetch,
    token: String,
    seen: HashSet<String>,
    pending: VecDeque<T>,
    limits: PaginationLimits,
    pages: usize,
    items: usize,
    finished: bool,
    failed: bool,
}

/// Creates a lazy paginator. `fetch_page` should call an ergonomic SDK list
/// method so identity, deadlines, retries, and response validation stay active.
pub fn paginate<T, Fetch>(
    fetch_page: Fetch,
    initial_page_token: impl Into<String>,
    limits: PaginationLimits,
) -> Paginator<T, Fetch> {
    let token = initial_page_token.into();
    let seen = if token.is_empty() {
        HashSet::new()
    } else {
        HashSet::from([token.clone()])
    };
    Paginator {
        fetch_page,
        token,
        seen,
        pending: VecDeque::new(),
        limits,
        pages: 0,
        items: 0,
        finished: false,
        failed: false,
    }
}

impl<T, Fetch> Paginator<T, Fetch> {
    /// Returns the next item, fetching pages only as needed.
    ///
    /// # Errors
    ///
    /// Returns the facade fetch error, a protocol error for a repeated token,
    /// or a non-retryable resource-exhausted error when a limit is reached.
    pub async fn try_next<FetchFuture>(&mut self) -> Result<Option<T>, Error>
    where
        Fetch: FnMut(String) -> FetchFuture,
        FetchFuture: Future<Output = Result<PaginationPage<T>, Error>>,
    {
        loop {
            if self.failed {
                return Ok(None);
            }
            if let Some(item) = self.pending.pop_front() {
                if self.items >= self.limits.max_items {
                    self.failed = true;
                    return Err(Error::pagination_limit(
                        "automatic pagination exceeded its item budget",
                    ));
                }
                self.items += 1;
                return Ok(Some(item));
            }
            if self.finished {
                return Ok(None);
            }
            if self.items >= self.limits.max_items {
                self.failed = true;
                return Err(Error::pagination_limit(
                    "automatic pagination exceeded its item budget",
                ));
            }
            if self.pages >= self.limits.max_pages {
                self.failed = true;
                return Err(Error::pagination_limit(
                    "automatic pagination exceeded its page budget",
                ));
            }
            let page = match (self.fetch_page)(self.token.clone()).await {
                Ok(page) => page,
                Err(error) => {
                    self.failed = true;
                    return Err(error);
                }
            };
            self.pages += 1;
            if !page.next_page_token.is_empty() && !self.seen.insert(page.next_page_token.clone()) {
                self.failed = true;
                return Err(Error::protocol(
                    "list response repeated an opaque page token",
                ));
            }
            self.finished = page.next_page_token.is_empty();
            self.token = page.next_page_token;
            self.pending = page.items.into();
        }
    }
}

/// Per-RPC behavior and correlation metadata. It does not redefine the wire
/// request.
#[derive(Clone, Debug, Default)]
pub struct CallOptions {
    request_id: Option<String>,
    trace_id: Option<String>,
    timeout: Option<Duration>,
    lease_token: Option<SensitiveLeaseToken>,
}

impl CallOptions {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Sets a caller-provided request identifier.
    ///
    /// # Errors
    ///
    /// Returns an error if the value is unsafe for gRPC metadata.
    pub fn with_request_id(mut self, value: impl Into<String>) -> Result<Self, Error> {
        let value = value.into();
        validate_metadata_value("request ID", &value, true)?;
        self.request_id = Some(value);
        Ok(self)
    }

    /// Sets the distributed trace identifier.
    ///
    /// # Errors
    ///
    /// Returns an error if the value is unsafe for gRPC metadata.
    pub fn with_trace_id(mut self, value: impl Into<String>) -> Result<Self, Error> {
        let value = value.into();
        validate_metadata_value("trace ID", &value, true)?;
        self.trace_id = Some(value);
        Ok(self)
    }

    /// Sets the overall RPC deadline, including retries.
    ///
    /// # Errors
    ///
    /// Returns an error for a zero or unbounded duration.
    pub fn with_timeout(mut self, value: Duration) -> Result<Self, Error> {
        if value.is_zero() || value > Duration::from_hours(24) {
            return Err(Error::invalid_argument(
                "call timeout must be positive and at most twenty-four hours",
            ));
        }
        self.timeout = Some(value);
        Ok(self)
    }

    /// Selects the raw scheduler-issued lease credential for a fenced worker
    /// command. Ordinary RPCs deliberately ignore this field; only a fenced
    /// facade can activate it through `prepare_fenced`.
    ///
    /// # Errors
    ///
    /// Returns an error unless the token is non-empty, bounded visible ASCII
    /// without whitespace or control characters.
    pub fn with_lease_token(mut self, value: impl Into<String>) -> Result<Self, Error> {
        let value = value.into();
        if value.is_empty()
            || value.len() > 4096
            || !value.bytes().all(|byte| (0x21..=0x7e).contains(&byte))
        {
            return Err(Error::invalid_argument(
                "lease token must contain at most 4096 visible non-whitespace ASCII characters",
            ));
        }
        self.lease_token = Some(SensitiveLeaseToken(value));
        Ok(self)
    }

    pub(crate) fn with_sensitive_lease_token(mut self, value: SensitiveLeaseToken) -> Self {
        self.lease_token = Some(value);
        self
    }

    pub(crate) fn prepare(&self, config: &Config) -> PreparedCall {
        let request_id = self.request_id.clone().unwrap_or_else(generate_request_id);
        let trace_id = self.trace_id.clone().unwrap_or_else(|| request_id.clone());
        let timeout = self.timeout.unwrap_or(config.rpc_timeout);
        PreparedCall {
            request_id,
            trace_id,
            deadline: Instant::now() + timeout,
            wall_deadline: SystemTime::now() + timeout,
            lease_token: None,
        }
    }

    pub(crate) fn prepare_fenced(&self, config: &Config) -> Result<PreparedCall, Error> {
        let mut prepared = self.prepare(config);
        prepared.lease_token = Some(
            self.lease_token
                .clone()
                .ok_or_else(|| Error::invalid_argument("fenced command requires a lease token"))?,
        );
        Ok(prepared)
    }

    pub(crate) fn bounded_by(&self, timeout: Duration) -> Self {
        let mut value = self.clone();
        value.timeout = Some(
            value
                .timeout
                .map_or(timeout, |existing| existing.min(timeout)),
        );
        value
    }
}

/// Behavior required for an idempotent training submission. The scientific
/// payload remains [`mindclade_protocols::training::v1::CreateTrainingRunCommand`].
#[derive(Clone, Debug)]
pub struct SubmitOptions {
    pub(crate) call: CallOptions,
    pub(crate) idempotency_key: String,
    pub(crate) correlation_id: String,
    pub(crate) causation_id: String,
    pub(crate) cancellation_token_id: String,
}

impl SubmitOptions {
    /// Creates validated behavior for an idempotent command.
    ///
    /// # Errors
    ///
    /// Returns an error when the key is empty or unsafe for transport
    /// metadata. The control plane computes the canonical command digest and
    /// materializes it before persistence; callers never reproduce
    /// language-specific protobuf serialization.
    pub fn new(idempotency_key: impl Into<String>) -> Result<Self, Error> {
        let idempotency_key = idempotency_key.into();
        validate_metadata_value("idempotency key", &idempotency_key, true)?;
        Ok(Self {
            call: CallOptions::default(),
            idempotency_key,
            correlation_id: String::new(),
            causation_id: String::new(),
            cancellation_token_id: String::new(),
        })
    }

    #[must_use]
    pub fn with_call_options(mut self, options: CallOptions) -> Self {
        self.call = options;
        self
    }

    /// Adds a correlation identifier.
    ///
    /// # Errors
    ///
    /// Returns an error if the identifier is unsafe for command context.
    pub fn with_correlation_id(mut self, value: impl Into<String>) -> Result<Self, Error> {
        self.correlation_id = checked_optional("correlation ID", value.into())?;
        Ok(self)
    }

    /// Adds a causation identifier.
    ///
    /// # Errors
    ///
    /// Returns an error if the identifier is unsafe for command context.
    pub fn with_causation_id(mut self, value: impl Into<String>) -> Result<Self, Error> {
        self.causation_id = checked_optional("causation ID", value.into())?;
        Ok(self)
    }

    /// Adds the durable cancellation-token identifier.
    ///
    /// # Errors
    ///
    /// Returns an error if the identifier is unsafe for command context.
    pub fn with_cancellation_token_id(mut self, value: impl Into<String>) -> Result<Self, Error> {
        self.cancellation_token_id = checked_optional("cancellation token ID", value.into())?;
        Ok(self)
    }
}

fn checked_optional(name: &str, value: String) -> Result<String, Error> {
    if !value.is_empty() {
        validate_metadata_value(name, &value, true)?;
    }
    Ok(value)
}

#[derive(Clone, Debug)]
pub(crate) struct PreparedCall {
    pub(crate) request_id: String,
    pub(crate) trace_id: String,
    pub(crate) deadline: Instant,
    pub(crate) wall_deadline: SystemTime,
    pub(crate) lease_token: Option<SensitiveLeaseToken>,
}

#[derive(Clone)]
pub(crate) struct SensitiveLeaseToken(String);

impl SensitiveLeaseToken {
    pub(crate) fn new(value: String) -> Result<Self, Error> {
        if value.len() < 32
            || value.len() > 4096
            || !value.bytes().all(|byte| (0x21..=0x7e).contains(&byte))
        {
            return Err(Error::protocol(
                "lease credential metadata must contain 32 through 4096 visible ASCII characters",
            ));
        }
        Ok(Self(value))
    }

    pub(crate) fn expose(&self) -> &str {
        &self.0
    }
}

impl fmt::Debug for SensitiveLeaseToken {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("[REDACTED]")
    }
}

impl PreparedCall {
    pub(crate) fn deadline_timestamp(&self) -> Result<Timestamp, Error> {
        let value = self
            .wall_deadline
            .duration_since(UNIX_EPOCH)
            .map_err(|_| Error::invalid_argument("call deadline precedes the Unix epoch"))?;
        let seconds = i64::try_from(value.as_secs())
            .map_err(|_| Error::invalid_argument("call deadline exceeds protobuf range"))?;
        let nanos = i32::try_from(value.subsec_nanos())
            .map_err(|_| Error::invalid_argument("call deadline nanos exceed protobuf range"))?;
        Ok(Timestamp { seconds, nanos })
    }

    pub(crate) fn command_context(
        &self,
        config: &Config,
        options: &SubmitOptions,
    ) -> Result<mindclade_protocols::common::v1::CommandContext, Error> {
        Ok(mindclade_protocols::common::v1::CommandContext {
            request_id: self.request_id.clone(),
            idempotency_key: options.idempotency_key.clone(),
            principal_id: config.identity.principal_id().to_owned(),
            trace_id: self.trace_id.clone(),
            deadline: Some(self.deadline_timestamp()?),
            // The server computes this over its received generated message,
            // then persists and emits the authoritative value.
            canonical_request_digest: String::new(),
            tenant_id: config.identity.tenant_id().to_owned(),
            project_id: config.identity.project_id().to_owned(),
            correlation_id: options.correlation_id.clone(),
            causation_id: options.causation_id.clone(),
            cancellation_token_id: options.cancellation_token_id.clone(),
        })
    }
}

pub(crate) fn validate_resource_value(name: &str, value: &str) -> Result<(), Error> {
    validate_metadata_value(name, value, true)
}

pub(crate) fn generate_request_id() -> String {
    let sequence = REQUEST_SEQUENCE.fetch_add(1, Ordering::Relaxed);
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |value| value.as_nanos());
    format!("{:x}-{nanos:x}-{sequence:x}", std::process::id())
}
